File: packages/core/src/nn/bitlinear.ts
import { PipelineManager } from "../gpu/pipeline.js";
import { BufferPool } from "../gpu/buffer-pool.js";
import { type BindGroupCache, createBGCache, clearBGCache, cachedBG } from "./bg-cache.js";
import type { ModelConfig } from "../types.js";

// Import shader sources (bundled as strings by tsup)
import rmsnormWGSL from "../shaders/rmsnorm.wgsl";
import quantizeWGSL from "../shaders/quantize.wgsl";
import rmsnormQuantizeWGSL from "../shaders/rmsnorm_quantize.wgsl";
import ternaryGemvWGSL from "../shaders/ternary_gemv.wgsl";
import ternaryGemmWGSL from "../shaders/ternary_gemm.wgsl";

/**
 * BitLinear layer: RMSNorm → Quantize → Ternary MatMul → Dequantize
 *
 * This is the core building block of BitNet. The weights are ternary {-1,0,+1}
 * packed as 2-bit values (16 per u32). Input activations are quantized to int8
 * with per-token absmax before the matmul.
 */
export class BitLinear {
  private device: GPUDevice;
  private pipelines: PipelineManager;
  private pool: BufferPool;

  // Weight buffers (on GPU)
  private packedWeights: GPUBuffer; // [outDim, inDim/16] u32
  private weightScales: GPUBuffer; // [outDim] f32
  private normWeight: GPUBuffer | null; // [inDim] f32 or null (skip RMSNorm)

  private inDim: number;
  private outDim: number;
  private kPacked: number;

  // Pre-created uniform buffers for N=1 decode (static params)
  private decodeNormUniform?: GPUBuffer;
  private decodeQuantUniform?: GPUBuffer;
  private decodeFusedUniform?: GPUBuffer;
  private decodeGemvParamsUniform?: GPUBuffer;
  private decodeGemvScaleUniform?: GPUBuffer;

  // Pre-created uniform buffers for N>1 prefill (dynamic — updated via writeBuffer)
  private prefillFusedUniform?: GPUBuffer;
  private prefillNormUniform?: GPUBuffer;
  private prefillQuantUniform?: GPUBuffer;
  private prefillGemmUniform?: GPUBuffer;

  // Bind group cache for N=1 decode
  private bgCache: BindGroupCache = createBGCache();

  constructor(
    device: GPUDevice,
    pipelines: PipelineManager,
    pool: BufferPool,
    packedWeights: GPUBuffer,
    weightScales: GPUBuffer,
    normWeight: GPUBuffer | null,
    inDim: number,
    outDim: number
  ) {
    this.device = device;
    this.pipelines = pipelines;
    this.pool = pool;
    this.packedWeights = packedWeights;
    this.weightScales = weightScales;
    this.normWeight = normWeight;
    this.inDim = inDim;
    this.outDim = outDim;
    this.kPacked = Math.ceil(inDim / 16);
  }

  /** Pre-create uniform buffers for the N=1 decode path (all static). */
  initDecodeUniforms(): void {
    if (this.normWeight) {
      // Fused RMSNorm+Quantize decode uniform
      const data = new ArrayBuffer(12);
      const v = new DataView(data);
      v.setUint32(0, 1, true);
      v.setUint32(4, this.inDim, true);
      v.setFloat32(8, 1e-5, true);
      this.decodeFusedUniform = this.createUniformBuffer(data);

      // Separate RMSNorm decode uniform (fallback)
      this.decodeNormUniform = this.createUniformBuffer(data);
    }
    {
      const data = new ArrayBuffer(8);
      const v = new DataView(data);
      v.setUint32(0, 1, true);
      v.setUint32(4, this.inDim, true);
      this.decodeQuantUniform = this.createUniformBuffer(data);
    }
    {
      const data = new ArrayBuffer(12);
      const v = new DataView(data);
      v.setUint32(0, this.outDim, true);
      v.setUint32(4, this.inDim, true);
      v.setUint32(8, this.kPacked, true);
      this.decodeGemvParamsUniform = this.createUniformBuffer(data);
    }
    this.decodeGemvScaleUniform = this.device.createBuffer({
      size: 4,
      usage: GPUBufferUsage.UNIFORM | GPUBufferUsage.COPY_DST,
    });

    // Prefill uniforms (reused via writeBuffer for N>1)
    const mkBuf = (size: number) =>
      this.device.createBuffer({
        size,
        usage: GPUBufferUsage.UNIFORM | GPUBufferUsage.COPY_DST,
      });
    if (this.normWeight) {
      this.prefillFusedUniform = mkBuf(12);
      this.prefillNormUniform = mkBuf(12);
    }
    this.prefillQuantUniform = mkBuf(8);
    this.prefillGemmUniform = mkBuf(16);
  }

  /**
   * Forward pass: input [N, inDim] f32 → output [N, outDim] f32
   * @param input Input buffer
   * @param N Number of tokens (sequence length)
   * @param encoder Command encoder to record dispatches
   */
  forward(
    input: GPUBuffer,
    N: number,
    encoder: GPUCommandEncoder
  ): GPUBuffer {
    // Step 1+2: RMSNorm + Quantize (fused when sub-norm exists, separate otherwise)
    let quantized: GPUBuffer;
    let inputScales: GPUBuffer;

    if (this.normWeight) {
      // Fused RMSNorm + Quantize: skip intermediate normed buffer
      quantized = this.pool.acquire(
        N * this.inDim * 4,
        GPUBufferUsage.STORAGE | GPUBufferUsage.COPY_SRC
      );
      inputScales = this.pool.acquire(
        N * 4,
        GPUBufferUsage.STORAGE | GPUBufferUsage.COPY_SRC | GPUBufferUsage.UNIFORM
      );
      this.dispatchFusedNormQuantize(encoder, input, quantized, inputScales, N);
    } else {
      // No sub-norm: just quantize
      quantized = this.pool.acquire(
        N * this.inDim * 4,
        GPUBufferUsage.STORAGE | GPUBufferUsage.COPY_SRC
      );
      inputScales = this.pool.acquire(
        N * 4,
        GPUBufferUsage.STORAGE | GPUBufferUsage.COPY_SRC | GPUBufferUsage.UNIFORM
      );
      this.dispatchQuantize(encoder, input, quantized, inputScales, N);
    }

    // Step 3: Ternary MatMul
    const output = this.pool.acquire(
      N * this.outDim * 4,
      GPUBufferUsage.STORAGE | GPUBufferUsage.COPY_SRC
    );

    if (N === 1) {
      this.dispatchGEMV(encoder, quantized, inputScales, output);
    } else {
      this.dispatchGEMM(encoder, quantized, inputScales, output, N);
    }

    // Release intermediates
    this.pool.release(quantized);
    this.pool.release(inputScales);

    return output;
  }

  private dispatchFusedNormQuantize(
    encoder: GPUCommandEncoder,
    input: GPUBuffer,
    output: GPUBuffer,
    scales: GPUBuffer,
    N: number
  ): void {
    const { pipeline, bindGroupLayout } = this.pipelines.getOrCreate(
      "rmsnorm_quantize",
      rmsnormQuantizeWGSL
    );

    let paramsBuffer: GPUBuffer;
    if (N === 1 && this.decodeFusedUniform) {
      paramsBuffer = this.decodeFusedUniform;
    } else if (this.prefillFusedUniform) {
      const paramsData = new ArrayBuffer(12);
      const v = new DataView(paramsData);
      v.setUint32(0, N, true);
      v.setUint32(4, this.inDim, true);
      v.setFloat32(8, 1e-5, true);
      this.device.queue.writeBuffer(this.prefillFusedUniform, 0, new Uint8Array(paramsData));
      paramsBuffer = this.prefillFusedUniform;
    } else {
      const paramsData = new ArrayBuffer(12);
      const v = new DataView(paramsData);
      v.setUint32(0, N, true);
      v.setUint32(4, this.inDim, true);
      v.setFloat32(8, 1e-5, true);
      paramsBuffer = this.createUniformBuffer(paramsData);
    }

    const entries: GPUBindGroupEntry[] = [
      { binding: 0, resource: { buffer: input } },
      { binding: 1, resource: { buffer: this.normWeight! } },
      { binding: 2, resource: { buffer: output } },
      { binding: 3, resource: { buffer: scales } },
      { binding: 4, resource: { buffer: paramsBuffer } },
    ];
    const bindGroup = N === 1
      ? cachedBG(this.bgCache, this.device, "fused_nq", bindGroupLayout, entries)
      : this.device.createBindGroup({ layout: bindGroupLayout, entries });

    const pass = encoder.beginComputePass();
    pass.setPipeline(pipeline);
    pass.setBindGroup(0, bindGroup);
    pass.dispatchWorkgroups(N);
    pass.end();
  }

  private dispatchRMSNorm(
    encoder: GPUCommandEncoder,
    input: GPUBuffer,
    output: GPUBuffer,
    N: number
  ): void {
    const { pipeline, bindGroupLayout } = this.pipelines.getOrCreate(
      "rmsnorm",
      rmsnormWGSL
    );

    let paramsBuffer: GPUBuffer;
    if (N === 1 && this.decodeNormUniform) {
      paramsBuffer = this.decodeNormUniform;
    } else if (this.prefillNormUniform) {
      const paramsData = new ArrayBuffer(12);
      const paramsView = new DataView(paramsData);
      paramsView.setUint32(0, N, true);
      paramsView.setUint32(4, this.inDim, true);
      paramsView.setFloat32(8, 1e-5, true);
      this.device.queue.writeBuffer(this.prefillNormUniform, 0, new Uint8Array(paramsData));
      paramsBuffer = this.prefillNormUniform;
    } else {
      const paramsData = new ArrayBuffer(12);
      const paramsView = new DataView(paramsData);
      paramsView.setUint32(0, N, true);
      paramsView.setUint32(4, this.inDim, true);
      paramsView.setFloat32(8, 1e-5, true);
      paramsBuffer = this.createUniformBuffer(paramsData);
    }

    const entries: GPUBindGroupEntry[] = [
      { binding: 0, resource: { buffer: input } },
      { binding: 1, resource: { buffer: this.normWeight! } },
      { binding: 2, resource: { buffer: output } },
      { binding: 3, resource: { buffer: paramsBuffer } },
    ];
    const bindGroup = N === 1
      ? cachedBG(this.bgCache, this.device, "rmsnorm", bindGroupLayout, entries)
      : this.device.createBindGroup({ layout: bindGroupLayout, entries });

    const pass = encoder.beginComputePass();
    pass.setPipeline(pipeline);
    pass.setBindGroup(0, bindGroup);
    pass.dispatchWorkgroups(N);
    pass.end();
  }

  private dispatchQuantize(
    encoder: GPUCommandEncoder,
    input: GPUBuffer,
    output: GPUBuffer,
    scales: GPUBuffer,
    N: number
  ): void {
    const { pipeline, bindGroupLayout } = this.pipelines.getOrCreate(
      "quantize",
      quantizeWGSL
    );

    let paramsBuffer: GPUBuffer;
    if (N === 1 && this.decodeQuantUniform) {
      paramsBuffer = this.decodeQuantUniform;
    } else if (this.prefillQuantUniform) {
      const paramsData = new ArrayBuffer(8);
      const paramsView = new DataView(paramsData);
      paramsView.setUint32(0, N, true);
      paramsView.setUint32(4, this.inDim, true);
      this.device.queue.writeBuffer(this.prefillQuantUniform, 0, new Uint8Array(paramsData));
      paramsBuffer = this.prefillQuantUniform;
    } else {
      const paramsData = new ArrayBuffer(8);
      const paramsView = new DataView(paramsData);
      paramsView.setUint32(0, N, true);
      paramsView.setUint32(4, this.inDim, true);
      paramsBuffer = this.createUniformBuffer(paramsData);
    }

    const entries: GPUBindGroupEntry[] = [
      { binding: 0, resource: { buffer: input } },
      { binding: 1, resource: { buffer: output } },
      { binding: 2, resource: { buffer: scales } },
      { binding: 3, resource: { buffer: paramsBuffer } },
    ];
    const bindGroup = N === 1
      ? cachedBG(this.bgCache, this.device, "quantize", bindGroupLayout, entries)
      : this.device.createBindGroup({ layout: bindGroupLayout, entries });

    const pass = encoder.beginComputePass();
    pass.setPipeline(pipeline);
    pass.setBindGroup(0, bindGroup);
    pass.dispatchWorkgroups(N);
    pass.end();
  }

  private dispatchGEMV(
    encoder: GPUCommandEncoder,
    input: GPUBuffer,
    inputScales: GPUBuffer,
    output: GPUBuffer
  ): void {
    const { pipeline, bindGroupLayout } = this.pipelines.getOrCreate(
      "ternary_gemv",
      ternaryGemvWGSL
    );

    let paramsBuffer: GPUBuffer;
    let inputScaleBuffer: GPUBuffer;
    if (this.decodeGemvParamsUniform && this.decodeGemvScaleUniform) {
      paramsBuffer = this.decodeGemvParamsUniform;
      inputScaleBuffer = this.decodeGemvScaleUniform;
    } else {
      const paramsData = new ArrayBuffer(12);
      const paramsView = new DataView(paramsData);
      paramsView.setUint32(0, this.outDim, true);
      paramsView.setUint32(4, this.inDim, true);
      paramsView.setUint32(8, this.kPacked, true);
      paramsBuffer = this.createUniformBuffer(paramsData);
      inputScaleBuffer = this.createUniformBuffer(new ArrayBuffer(4));
    }
    // Copy from inputScales[0] to uniform
    encoder.copyBufferToBuffer(inputScales, 0, inputScaleBuffer, 0, 4);

    const entries: GPUBindGroupEntry[] = [
      { binding: 0, resource: { buffer: this.packedWeights } },
      { binding: 1, resource: { buffer: input } },
      { binding: 2, resource: { buffer: this.weightScales } },
      { binding: 3, resource: { buffer: paramsBuffer } },
      { binding: 4, resource: { buffer: inputScaleBuffer } },
      { binding: 5, resource: { buffer: output } },
    ];
    const bindGroup = cachedBG(this.bgCache, this.device, "gemv", bindGroupLayout, entries);

    const pass = encoder.beginComputePass();
    pass.setPipeline(pipeline);
    pass.setBindGroup(0, bindGroup);
    pass.dispatchWorkgroups(this.outDim);
    pass.end();
  }

  private dispatchGEMM(
    encoder: GPUCommandEncoder,
    input: GPUBuffer,
    inputScales: GPUBuffer,
    output: GPUBuffer,
    N: number
  ): void {
    const { pipeline, bindGroupLayout } = this.pipelines.getOrCreate(
      "ternary_gemm",
      ternaryGemmWGSL
    );

    // Params: M, N, K, K_packed
    const paramsData = new ArrayBuffer(16);
    const paramsView = new DataView(paramsData);
    paramsView.setUint32(0, this.outDim, true);
    paramsView.setUint32(4, N, true);
    paramsView.setUint32(8, this.inDim, true);
    paramsView.setUint32(12, this.kPacked, true);

    let paramsBuffer: GPUBuffer;
    if (this.prefillGemmUniform) {
      this.device.queue.writeBuffer(this.prefillGemmUniform, 0, new Uint8Array(paramsData));
      paramsBuffer = this.prefillGemmUniform;
    } else {
      paramsBuffer = this.createUniformBuffer(paramsData);
    }

    const bindGroup = this.device.createBindGroup({
      layout: bindGroupLayout,
      entries: [
        { binding: 0, resource: { buffer: this.packedWeights } },
        { binding: 1, resource: { buffer: input } },
        { binding: 2, resource: { buffer: this.weightScales } },
        { binding: 3, resource: { buffer: paramsBuffer } },
        { binding: 4, resource: { buffer: inputScales } },
        { binding: 5, resource: { buffer: output } },
      ],
    });

    const wgM = Math.ceil(this.outDim / 64);
    const wgN = Math.ceil(N / 64);

    const pass = encoder.beginComputePass();
    pass.setPipeline(pipeline);
    pass.setBindGroup(0, bindGroup);
    pass.dispatchWorkgroups(wgM, wgN);
    pass.end();
  }

  /** Clear the bind group cache (call on KV cache reset). */
  clearBGCache(): void {
    clearBGCache(this.bgCache);
  }

  private createUniformBuffer(data: ArrayBuffer): GPUBuffer {
    const size = Math.max(Math.ceil(data.byteLength / 4) * 4, 4);
    const buffer = this.device.createBuffer({
      size,
      usage: GPUBufferUsage.UNIFORM | GPUBufferUsage.COPY_DST,
      mappedAtCreation: true,
    });
    new Uint8Array(buffer.getMappedRange()).set(new Uint8Array(data));
    buffer.unmap();
    return buffer;
  }
}

File: packages/rust/crates/oxbitnet/src/nn/bitlinear.rs
use std::sync::Arc;

use wgpu::BufferUsages;

use crate::gpu::buffer_pool::{BufferPool, GpuBuf};
use crate::gpu::pipeline::PipelineManager;
use crate::nn::bg_cache::BgCache;

const RMSNORM_WGSL: &str = include_str!("../shaders/rmsnorm.wgsl");
const QUANTIZE_WGSL: &str = include_str!("../shaders/quantize.wgsl");
const TERNARY_GEMV_WGSL: &str = include_str!("../shaders/ternary_gemv.wgsl");
const TERNARY_GEMM_WGSL: &str = include_str!("../shaders/ternary_gemm.wgsl");

/// BitLinear layer: RMSNorm → Quantize → Ternary MatMul → Dequantize
///
/// Core building block of BitNet. Weights are ternary {-1,0,+1} packed as
/// 2-bit values (16 per u32). Input activations are quantized to int8
/// with per-token absmax before the matmul.
pub struct BitLinear {
    device: Arc<wgpu::Device>,
    packed_weights: GpuBuf,
    weight_scales: GpuBuf,
    norm_weight: Option<GpuBuf>,
    pub(crate) in_dim: usize,
    pub(crate) out_dim: usize,
    k_packed: usize,
    bg_cache: BgCache,
}

impl BitLinear {
    pub fn new(
        device: Arc<wgpu::Device>,
        packed_weights: GpuBuf,
        weight_scales: GpuBuf,
        norm_weight: Option<GpuBuf>,
        in_dim: usize,
        out_dim: usize,
    ) -> Self {
        Self {
            device,
            packed_weights,
            weight_scales,
            norm_weight,
            in_dim,
            out_dim,
            k_packed: (in_dim + 15) / 16,
            bg_cache: BgCache::new(),
        }
    }

    /// Forward pass: input [N, in_dim] f32 → output [N, out_dim] f32
    pub fn forward(
        &mut self,
        input: &GpuBuf,
        n: usize,
        encoder: &mut wgpu::CommandEncoder,
        pipelines: &mut PipelineManager,
        pool: &BufferPool,
    ) -> GpuBuf {
        // Step 1: RMSNorm (optional)
        let normed = if let Some(ref _norm_w) = self.norm_weight {
            let out = pool.acquire(
                (n * self.in_dim * 4) as u64,
                BufferUsages::STORAGE | BufferUsages::COPY_SRC,
            );
            self.dispatch_rmsnorm(encoder, input, &out, n, pipelines);
            Some(out)
        } else {
            None
        };
        let normed_ref = normed.as_ref().unwrap_or(input);

        // Step 2: Quantize (absmax int8)
        let quantized = pool.acquire(
            (n * self.in_dim * 4) as u64,
            BufferUsages::STORAGE | BufferUsages::COPY_SRC,
        );
        let input_scales = pool.acquire(
            (n * 4) as u64,
            BufferUsages::STORAGE | BufferUsages::COPY_SRC | BufferUsages::UNIFORM,
        );
        self.dispatch_quantize(encoder, normed_ref, &quantized, &input_scales, n, pipelines);

        // Step 3: Ternary MatMul
        let output = pool.acquire(
            (n * self.out_dim * 4) as u64,
            BufferUsages::STORAGE | BufferUsages::COPY_SRC,
        );

        if n == 1 {
            self.dispatch_gemv(encoder, &quantized, &input_scales, &output, pipelines);
        } else {
            self.dispatch_gemm(encoder, &quantized, &input_scales, &output, n, pipelines);
        }

        output
    }

    fn dispatch_rmsnorm(
        &mut self,
        encoder: &mut wgpu::CommandEncoder,
        input: &wgpu::Buffer,
        output: &wgpu::Buffer,
        n: usize,
        pipelines: &mut PipelineManager,
    ) {
        let entry = pipelines.get_or_create_default("rmsnorm", RMSNORM_WGSL);
        let params = create_uniform_u32_u32_f32(
            &self.device,
            n as u32,
            self.in_dim as u32,
            1e-5,
        );

        let bg = self.device.create_bind_group(&wgpu::BindGroupDescriptor {
            label: Some("bitlinear_rmsnorm"),
            layout: &entry.bind_group_layout,
            entries: &[
                buf_entry(0, input),
                buf_entry(1, self.norm_weight.as_ref().unwrap()),
                buf_entry(2, output),
                buf_entry(3, &params),
            ],
        });

        let mut pass = encoder.begin_compute_pass(&Default::default());
        pass.set_pipeline(&entry.pipeline);
        pass.set_bind_group(0, Some(&bg), &[]);
        pass.dispatch_workgroups(n as u32, 1, 1);
    }

    fn dispatch_quantize(
        &self,
        encoder: &mut wgpu::CommandEncoder,
        input: &wgpu::Buffer,
        output: &wgpu::Buffer,
        scales: &wgpu::Buffer,
        n: usize,
        pipelines: &mut PipelineManager,
    ) {
        let entry = pipelines.get_or_create_default("quantize", QUANTIZE_WGSL);
        let params = create_uniform_u32_u32(&self.device, n as u32, self.in_dim as u32);

        let bg = self.device.create_bind_group(&wgpu::BindGroupDescriptor {
            label: Some("bitlinear_quantize"),
            layout: &entry.bind_group_layout,
            entries: &[
                buf_entry(0, input),
                buf_entry(1, output),
                buf_entry(2, scales),
                buf_entry(3, &params),
            ],
        });

        let mut pass = encoder.begin_compute_pass(&Default::default());
        pass.set_pipeline(&entry.pipeline);
        pass.set_bind_group(0, Some(&bg), &[]);
        pass.dispatch_workgroups(n as u32, 1, 1);
    }

    fn dispatch_gemv(
        &self,
        encoder: &mut wgpu::CommandEncoder,
        input: &wgpu::Buffer,
        input_scales: &wgpu::Buffer,
        output: &wgpu::Buffer,
        pipelines: &mut PipelineManager,
    ) {
        let entry = pipelines.get_or_create_default("ternary_gemv", TERNARY_GEMV_WGSL);

        let params_data = [
            (self.out_dim as u32).to_le_bytes(),
            (self.in_dim as u32).to_le_bytes(),
            (self.k_packed as u32).to_le_bytes(),
        ]
        .concat();
        let params_buf = create_uniform_raw(&self.device, &params_data);

        // Copy input_scales[0] to a uniform buffer
        let scale_uniform = self.device.create_buffer(&wgpu::BufferDescriptor {
            label: Some("gemv_scale"),
            size: 4,
            usage: BufferUsages::UNIFORM | BufferUsages::COPY_DST,
            mapped_at_creation: false,
        });
        encoder.copy_buffer_to_buffer(input_scales, 0, &scale_uniform, 0, 4);

        let bg = self.device.create_bind_group(&wgpu::BindGroupDescriptor {
            label: Some("bitlinear_gemv"),
            layout: &entry.bind_group_layout,
            entries: &[
                buf_entry(0, &self.packed_weights),
                buf_entry(1, input),
                buf_entry(2, &self.weight_scales),
                buf_entry(3, &params_buf),
                buf_entry(4, &scale_uniform),
                buf_entry(5, output),
            ],
        });

        let mut pass = encoder.begin_compute_pass(&Default::default());
        pass.set_pipeline(&entry.pipeline);
        pass.set_bind_group(0, Some(&bg), &[]);
        pass.dispatch_workgroups(self.out_dim as u32, 1, 1);
    }

    fn dispatch_gemm(
        &self,
        encoder: &mut wgpu::CommandEncoder,
        input: &wgpu::Buffer,
        input_scales: &wgpu::Buffer,
        output: &wgpu::Buffer,
        n: usize,
        pipelines: &mut PipelineManager,
    ) {
        let entry = pipelines.get_or_create_default("ternary_gemm", TERNARY_GEMM_WGSL);

        let params_data = [
            (self.out_dim as u32).to_le_bytes(),
            (n as u32).to_le_bytes(),
            (self.in_dim as u32).to_le_bytes(),
            (self.k_packed as u32).to_le_bytes(),
        ]
        .concat();
        let params_buf = create_uniform_raw(&self.device, &params_data);

        let bg = self.device.create_bind_group(&wgpu::BindGroupDescriptor {
            label: Some("bitlinear_gemm"),
            layout: &entry.bind_group_layout,
            entries: &[
                buf_entry(0, &self.packed_weights),
                buf_entry(1, input),
                buf_entry(2, &self.weight_scales),
                buf_entry(3, &params_buf),
                buf_entry(4, input_scales),
                buf_entry(5, output),
            ],
        });

        let wg_m = ((self.out_dim + 63) / 64) as u32;
        let wg_n = ((n + 63) / 64) as u32;

        let mut pass = encoder.begin_compute_pass(&Default::default());
        pass.set_pipeline(&entry.pipeline);
        pass.set_bind_group(0, Some(&bg), &[]);
        pass.dispatch_workgroups(wg_m, wg_n, 1);
    }

    pub fn clear_bg_cache(&mut self) {
        self.bg_cache.clear();
    }
}

// --- Uniform buffer helpers ---

pub(crate) fn buf_entry(binding: u32, buffer: &wgpu::Buffer) -> wgpu::BindGroupEntry<'_> {
    wgpu::BindGroupEntry {
        binding,
        resource: buffer.as_entire_binding(),
    }
}

pub(crate) fn create_uniform_raw(device: &wgpu::Device, data: &[u8]) -> wgpu::Buffer {
    let size = ((data.len().max(4) + 3) / 4 * 4) as u64;
    let buffer = device.create_buffer(&wgpu::BufferDescriptor {
        label: None,
        size,
        usage: BufferUsages::UNIFORM | BufferUsages::COPY_DST,
        mapped_at_creation: true,
    });
    {
        let mut view = buffer.slice(..).get_mapped_range_mut();
        view[..data.len()].copy_from_slice(data);
    }
    buffer.unmap();
    buffer
}

pub(crate) fn create_uniform_u32_u32(device: &wgpu::Device, a: u32, b: u32) -> wgpu::Buffer {
    let data = [a.to_le_bytes(), b.to_le_bytes()].concat();
    create_uniform_raw(device, &data)
}

pub(crate) fn create_uniform_u32_u32_f32(
    device: &wgpu::Device,
    a: u32,
    b: u32,
    c: f32,
) -> wgpu::Buffer {
    let data = [a.to_le_bytes(), b.to_le_bytes(), c.to_le_bytes()].concat();
    create_uniform_raw(device, &data)
}
